quinn-proto/src/config.rs

```diff
@@ -525,14 +525,38 @@ where
 
 #[cfg(feature = "rustls")]
 impl ServerConfig<crypto::rustls::TlsSession> {
-    /// Set the certificate chain that will be presented to clients
-    pub fn certificate(
-        &mut self,
+    /// Create a server config with the given certificate chain to be presented to clients
+    ///
+    /// Uses a randomized handshake token key.
+    pub fn with_single_cert(
         cert_chain: CertificateChain,
         key: PrivateKey,
-    ) -> Result<&mut Self, rustls::TLSError> {
-        Arc::make_mut(&mut self.crypto).set_single_cert(cert_chain.certs, key.inner)?;
-        Ok(self)
+    ) -> Result<Self, rustls::TLSError> {
+        let mut crypto = rustls::ServerConfig::with_ciphersuites(
+            rustls::NoClientAuth::new(),
+            &crypto::rustls::QUIC_CIPHER_SUITES,
+        );
+
+        crypto.versions = vec![rustls::ProtocolVersion::TLSv1_3];
+        crypto.set_single_cert(cert_chain.certs, key.inner)?;
+        crypto.max_early_data_size = u32::max_value();
+
+        Ok(Self::with_crypto(Arc::new(crypto)))
+    }
+
+    /// Create a server config with the given [`rustls::ServerConfig`]
+    ///
+    /// Uses a randomized handshake token key.
+    pub fn with_crypto(crypto: Arc<rustls::ServerConfig>) -> Self {
```

Comment:
Do you want to validate here that the `ServerConfig` is suitable for QUIC? E.g. check for TLS1.3 only protocol and QUIC cipher suites?

Reply:
Right now `ServerConfig` does not support introspection of cipher suites and protocol versions. There are no QUIC cipher suites AFAICT, there are TLS 1.3 cipher suites and all of these are supported by QUIC (although Rustls does not support the CCM cipher suites). The config builder pattern in Rustls 0.20 makes sure that protocol versions and cipher suites configured in Rustls match; it also returns an error for new QUIC connections if the configuration cannot support QUIC. We can add a check here once https://github.com/rustls/rustls/pull/834 is released.